metrics/src/main/java/org/aerogear/mobile/metrics/providers/DefaultMetricsProvider.java

```diff
@@ -0,0 +1,82 @@
+package org.aerogear.mobile.metrics.providers;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+import android.content.pm.PackageManager;
+import android.os.Build;
+
+import org.aerogear.mobile.core.MobileCore;
+import org.aerogear.mobile.metrics.MetricsService;
+import org.aerogear.mobile.metrics.interfaces.MetricsProvider;
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.UUID;
+
+public class DefaultMetricsProvider extends MetricsProvider {
+    public final static String STORAGE_NAME = "org.aerogear.mobile.metrics";
+    public final static String STORAGE_KEY = "metrics-sdk-installation-id";
+
+    /**
+     * Get or create the client ID that identifies a device as long as the user doesn't
+     * reinstall the app or delete the app storage. A random UUID is created and stored in the
+     * application shared preferences.
+     *
+     * Can be overridden to provide a different implementation for identification.
+     *
+     * @param context Android app context
+     * @return String Client ID
+     */
+    protected String getOrCreateClientId(final Context context) {
+        final SharedPreferences preferences = context
+            .getSharedPreferences(STORAGE_NAME, Context.MODE_PRIVATE);
+
+        String clientId = preferences.getString(STORAGE_KEY, null);
+        if (clientId == null) {
+            clientId = UUID.randomUUID().toString();
+
+            MobileCore.getLogger().info(MetricsService.TAG, "Generated a new client ID: " + clientId);
+
+            SharedPreferences.Editor editor = preferences.edit();
+            editor.putString(STORAGE_KEY, clientId);
+            editor.commit();
+        }
+
+        return clientId;
+    }
+
+    /**
+     * Get the user app version from the package manager
+     *
+     * @param context Android application context
+     * @return String app version name
+     */
+    private String getAppVersion(final Context context) throws JSONException {
+        try {
+            return context
+                .getPackageManager()
+                .getPackageInfo(context.getPackageName(), 0)
+                .versionName;
+        } catch (PackageManager.NameNotFoundException e) {
+            // Wrap in Initialization exception
+            throw new JSONException(e.getMessage());
+        }
+    }
+
+    @Override
+    public String namespace() {
+        return null;
```

Comment:
If namespace is null in the abstract and concrete classes, why do we have it?

Reply:
In that case `null` has a specific meaning: the properties will be added to the root of the metrics object. You need to override it and return a value if you want namespacing.

Reply:
Ah sorry, minunderstood the comment. Yes this can be removed.